src/python/pants/backend/url_handlers/s3/integration_test.py

```diff
@@ -197,3 +222,33 @@ def send_headers(self):
         )
     assert snapshot.files == ("file.txt",)
     assert snapshot.digest == DOWNLOADS_EXPECTED_DIRECTORY_DIGEST
+
+
+def test_aws_credentials_caching(rule_runner: RuleRunner, monkeypatch_botocore) -> None:
+    """Test that AWS credentials are properly cached based on environment variables."""
+    monkeypatch_botocore("https://example.com")
+
+    def set_aws_env_vars(rule_runner: RuleRunner, env_vars: dict[str, str]) -> None:
+        rule_runner.set_options(
+            args=[],
+            env=env_vars,
+        )
+
+    set_aws_env_vars(rule_runner, {"AWS_PROFILE": "profile1"})
+
+    creds1 = rule_runner.request(AWSCredentials, [])
+    creds2 = rule_runner.request(AWSCredentials, [])
+    assert creds1 is creds2
+    assert creds1.creds.access_key == "ACCESS_profile1"
+
+    # Request with different environment should return different credentials
+    set_aws_env_vars(rule_runner, {"AWS_PROFILE": "profile2"})
+    creds3 = rule_runner.request(AWSCredentials, [])
+    assert creds1 is not creds3
+    assert creds3.creds.access_key == "ACCESS_profile2"
+
+    # Request with original environment should return original credentials
+    set_aws_env_vars(rule_runner, {"AWS_PROFILE": "profile1"})
+    creds4 = rule_runner.request(AWSCredentials, [])
+    # N.B. Not totally sure why, but 'is' doesn't work here because of how set_options operates
+    assert creds1 == creds4
```

Comment:
I'm not totally happy with these tests as I couldn't get the assertions working that the _exact same_ object is returned, but in local testing with pantsd it appears to work correctly